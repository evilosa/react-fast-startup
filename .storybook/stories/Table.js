import React from 'react';
import { withInfo } from '@storybook/addon-info';
import { storiesOf } from '@storybook/react';
import Table from '../../src/Table';
import TableColumn from '../../src/TableColumn';

storiesOf('Table', module)
  .add('default',
    withInfo({
      components: { Table },
      header: true,
      inline: true,
      text: 'Table component',
    })(() => {
      const items = [
          { id: '1', title: 'Toast', sum: '$5'},
          { id: '2', title: 'Chicken', sum: '$15'},
          { id: '3', title: 'Coffee', sum: '$3'},
        ];
      return (
        <div style={{height: '900px', display: 'flex'}}>
          <Table
            header="Test table"
            items={items}
          >
            <TableColumn header="Id" propName="id"/>
            <TableColumn header="Title" propName="title"/>
            <TableColumn header="Sum" propName="sum"/>
          </Table>
        </div>
      )
    }));