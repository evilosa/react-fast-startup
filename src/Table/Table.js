import * as React from 'react';
import PropTypes from 'prop-types';

class Table extends React.Component {

  static propTypes = {
    items: PropTypes.array.isRequired,
    canAddNewItem: PropTypes.bool,
    children: PropTypes.node,
  };

  static defaultProps = {
    items: [],
    canAddNewItem: false,
    header: '',
  };

  constructor(props) {
    super(props);

    const { items } = props;

    this.state = {
      items,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.items !== this.props.items) {
      this.setState(prev => ({
        ...prev,
        items: nextProps.items,
      }));
    }
  }

  _renderHeaders = (headerColumns) => {
    return headerColumns.map(({header, width}, index) => (
      <div className="table-header-column" key={index}>{header}</div>
    ));
  };

  render() {
    const { children, header, items } = this.props;
    const headerColumns = [];
    const bodyColumns = [];
    const footerColumns = [];

    React.Children.forEach(children, (child) => {
      if (!child.type || child.type.displayName !== 'TableColumn')
        throw new Error('Child component should be instance of <TableColumn />');

      const { header, propName, footer, width } = child.props;

      // Fill headers columns
      headerColumns.push({ header, width });

      // Fill body columns
      bodyColumns.push({ propName, width });

      // Fill footer
      footerColumns.push({ footer, width });
    });

    return (
      <div>
        <h1>{header}</h1>
        <div className="table-header">
          {this._renderHeaders(headerColumns)}
        </div>
        <div className="table-body"></div>
        <div className="table-footer"></div>
      </div>
    );
  }
};

export default Table;