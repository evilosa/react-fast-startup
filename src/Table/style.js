export const TableColors = {
  baseColor: '#fff',
  baseBackground: '#222222',
  primaryBackground: '#0074d9',
  warningBackground: '#f00',
};

export const createTableStyle = (colors = TableColors) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  header: {
    display: 'flex',
    flex: '1 0',
  },

  body: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  footer: {
    display: 'flex',
    flex: '1 0',
  },

  tableRow: {
    display: 'flex',
    flex: '1 0',
  },

  tableHeaderColumn: {
    display: 'flex',
    flex: '1 0',
    justifyContent: 'flex-start',
    alignItems: 'center',
    fontWeight: 'bold'
  },

  tableRowColumn: {
    flex: '1 0',
  }
});

export const TableStyle = createTableStyle(TableColors);

export default TableStyle;
