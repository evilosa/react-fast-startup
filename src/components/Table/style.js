export const TableColors = {
  baseBackground: '#fff',
  evenBackground: '#EEF3F5',
  border: '#D7DBDD',
  headerBackground: '#FCFDFD',
  headerText: '#16A2D7'
}

export const createTableStyle = (colors = TableColors) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  header: {
    display: 'flex',
    flex: '1 0',
    backgroundColor: colors.headerBackground,
    minHeight: '30px',
    maxHeight: '30px',
    borderBottomWidth: '1px',
    borderBottomStyle: 'solid',
    borderBottomColor: colors.border
  },

  body: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  footer: {
    display: 'flex',
    flex: '1 0',
    minHeight: '30px',
    maxHeight: '30px',
  },

  tableRow: {
    display: 'flex',
    flex: '1 0',
    background: colors.baseBackground,
    minHeight: '30px',
    maxHeight: '30px',
  },

  even: {
    background: colors.evenBackground,
  },

  tableHeaderColumn: {
    display: 'flex',
    flex: '1 0',
    padding: '5px',
    justifyContent: 'flex-start',
    alignItems: 'center',
    fontFamily: "'Raleway', sans-serif",
    fontSize: '12px',
    fontWeight: '400',
    lineHeight: '18px',
    fontStretch: '100%',
    color: colors.headerText,
  },

  tableRowActionColumn: {
    flex: '1 0',
    padding: '5px',
    fontFamily: "'Raleway', sans-serif",
    fontSize: '12px',
    fontWeight: '400',
    lineHeight: '18px',
    fontStretch: '100%',
  },

  tableRowColumn: {
    flex: '1 0',
    padding: '5px',
    fontFamily: "'Raleway', sans-serif",
    fontSize: '12px',
    fontWeight: '400',
    lineHeight: '18px',
    fontStretch: '100%',
  }
})

export const TableStyle = createTableStyle(TableColors)

export default TableStyle
