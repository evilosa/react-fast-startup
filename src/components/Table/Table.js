import * as React from 'react'
import Radium from 'radium'
import PropTypes from 'prop-types'
import style from './style'

class Table extends React.Component {

  static propTypes = {
    items: PropTypes.array.isRequired,
    canAddNewItem: PropTypes.bool,
    children: PropTypes.node,
  };

  static defaultProps = {
    items: [],
    canAddNewItem: false,
    header: '',
  };

  constructor(props) {
    super(props);

    const { items } = props;

    this.state = {
      items,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.items !== this.props.items) {
      this.setState(prev => ({
        ...prev,
        items: nextProps.items,
      }));
    }
  }

  _renderHeaders = (headerColumns) => {
    return headerColumns.map(({header, width}, index) => (
      <div className="table-header-column" key={index} style={[style.tableHeaderColumn, { flex: width }]}>{header}</div>
    ));
  };

  _renderRows = (rowColumns) => {
    const { items } = this.state;
    return items.map((rowData, rowIndex) => (
      <div className="table-row" key={rowIndex} style={[style.tableRow, rowIndex % 2 === 0 && style.even]}>
        {this._renderRowColumns(rowColumns, rowData)}
      </div>
    ));
  };

  _renderRowColumns = (rowColumns, rowData) => (
    rowColumns.map((column, index) => {
      const columnData = rowData[column.propName];
      return <div className="table-row-column" key={index} style={style.tableRowColumn}>{columnData}</div>
    })
  );

  render() {
    const { children, header } = this.props;
    const headerColumns = [];
    const rowColumns = [];
    const footerColumns = [];

    React.Children.forEach(children, (child) => {
      if (!child.type || child.type.displayName !== 'TableColumn')
        throw new Error('Child component should be instance of <TableColumn />');

      const { header, propName, footer, width } = child.props;

      // Fill headers columns
      headerColumns.push({ header, width });

      // Fill row columns
      rowColumns.push({ propName, width });

      // Fill footer
      footerColumns.push({ footer, width });
    });

    return (
      <div style={style.root}>
        <h1>{header}</h1>

        {/* Header */}
        <div
          className="table-header"
          style={style.header}
        >
          {this._renderHeaders(headerColumns)}
        </div>

        {/* Body */}
        <div
          className="table-body"
          style={style.body}
        >
          {rowColumns.length > 0 && this._renderRows(rowColumns)}
        </div>

        {/* Footer */}
        <div
          className="table-footer"
          style={style.footer}
        >
        </div>
      </div>
    );
  }
};

export default Radium(Table);