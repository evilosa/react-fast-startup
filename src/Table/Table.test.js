import React from 'react';
import { shallow, mount } from 'enzyme';
import Table from './Table';
// import TableColumn from '../TableColumn';

describe('Table', () => {
  it('renders properly', () => {
    const table = shallow(<Table/>);
    expect(table).toMatchSnapshot();
  });

  describe('props ', () => {
    const table = mount(<Table/>);
    const props = table.props();

    it('initialized with items', () => {
      expect(props.items).toEqual([]);
    });

    it('initialized with canAddNewItem option', () => {
      expect(props.canAddNewItem).toBe(false);
    });

    it('initialized with empty header', () => {
      expect(props.header).toEqual('');
    });

    it('initialize default items state and update state with new items prop', () => {
      const testItem = { id: '1', name: 'Test name' };
      expect(table.state().items).toEqual([]);
      table.setProps({items: [ testItem ]});
      expect(table.state().items).toEqual([ testItem ]);
    });
  });

});