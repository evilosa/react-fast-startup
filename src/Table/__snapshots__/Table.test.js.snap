// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders properly 1`] = `
<div>
  <h1 />
  <div
    className="table-header"
  />
  <div
    className="table-body"
  />
  <div
    className="table-footer"
  />
</div>
`;
